src/findLinkRange.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import Range from '../engine/model/range.js';
+import Position from '../engine/model/position.js';
+
+/**
+ * Walk backward and forward from start position node by node as long as they have the same `linkHref` attribute value and return
+ * {@link engine.model.Range Range} with founded link.
+ *
+ * @param {engine.model.Position} position Start position.
+ * @param {String} value `linkHref` attribute value.
+ * @returns {engine.model.Range} Link range.
+ */
+export default function findLinkRange( position, value ) {
+	return new Range( _findBackwardBound( position, value ), _findForwardBound( position, value ) );
+}
+
+// Walk backward node by node as long as they have the same `linkHref` attribute value and return position just before last matched node.
+//
+// @param {engine.model.Position} position Start position.
+// @param {String} value `linkHref` attribute value.
+// @returns {engine.model.Position} Position just before last matched node.
+function _findBackwardBound( position, value ) {
+	// Get node before position. When position is inside text node then start searching from text node.
+	let node = position.textNode === null ? position.nodeBefore : position.textNode;
+
+	// When node doesn't match it means we are at the beginning of link.
+	if ( !node || node.getAttribute( 'linkHref' ) != value ) {
+		return position;
+	}
+
+	let lastNode = node;
+
+	while ( node ) {
+		// When node doesn't match it means that previous node is the last matches node and we need to return position before it.
+		if ( node.getAttribute( 'linkHref' ) != value ) {
+			return Position.createBefore( lastNode );
+		}
+
+		lastNode = node;
+		node = node.previousSibling;
+	}
+
+	// We reach the bound.
+	return Position.createBefore( lastNode );
+}
```

Comment:
I think it could be shorter and in one function. You should add `lookBack` or `directionBackward` flag (`Boolean`) and rename it to `_findBound()`.

Then you could either modify your function to use this flag, i.e.:
`node = node.previousSibling;` -> `node = lookBack ? node.previousSibling : node.nextSibling;`

Or use one of these (I am not sure if these work, I just wrote them down because I had feeling that this part of code can be done in less LOC and in one function):

```javascript
function _findBound( position, value, lookBack ) {
	let node = position.textNode || ( lookBack ? position.nodeBefore : position.nodeAfter );

	if ( !node ) {
		return position;
	}

	let lastNode = node;

	while ( node && node.getAttribute( 'linkHref' ) == value ) {
		lastNode = node;
		node = lookBack ? node.previousSibling : node.nextSibling;
	}

	return Position.createAt( lastNode, lookBack ? 'before' : 'after' );
}

return new Range( _findBound( position, value, true ), _findBound( position, value, false ) );
```
```javascript
function _findBound( position, value, lookBack ) {
	let lastNode = node;

	while ( node && node.getAttribute( 'linkHref' ) == value ) {
		lastNode = node;
		node = lookBack ? node.previousSibling : node.nextSibling;
	}

	return Position.createAt( lastNode, lookBack ? 'before' : 'after' );
}

const prevBound = position.textNode || position.nodeBefore ? _findBound( position, value, true ) : position;
const nextBound = position.textNode || position.nodeAfter ? _findBound( position, value, false ) : position;

return new Range( prevBound, nextBound );
```